### DISCUSSION TITLE
Using ellipsis in a TypeVarTuple

### DISCUSSION THREAD

**Discussion by MilanStaffehl** [python/typing#1790] [Q&A] [answered]
I want to have an alias for `tuple`, but I wish to keep it general purpose, so I am parametrizing it with a `TypeVarTuple`:

```Python
from typing import TypeVarTuple

_ShapeTypeVar = TypeVarTuple('_ShapeTypeVar')
Shape = tuple[*_ShapeTypeVar]

x: Shape[int, ...]
```

This works fine for all types (i.e. I can use this type like `Shape[int, str, float]`), but does not work for ellipses. If I use my alias with an ellipsis, all type checkers give me an error, stating that this is not supported behavior:

```
C:\Desktop> mypy tmp.py
tmp.py:6: error: Unexpected "..."  [misc]

C:\Desktop> pyright tmp.py
C:\Desktop\tmp.py:6:15 - error: "..." is not allowed in this context
```

So, either I am not understanding the typing system correctly, or there is something not quite right. I want to make sure I am not doing it completely wrong, so any guidance would be appreciated.

Some things I have tried:

- Using `typing.Tuple` yields the same result.
- Using `class Shape[tuple[*_ShapeTypeVar]]` yields the same result.
- Using `Shape[*Tuple[int, ..]]` or `Shape[*tuple[int, ...]]` works, but is somewhat unsatisfying and unintuitive.

**Is this a valid way to create an alias for `tuple` or am I doing it wrong?** Are ellipses just not valid in a `TypeVarTuple`? If so, what is the correct way to get a parametrized alias of `tuple` that supports ellipses?

**Comment by erictraut** [comment, marked as answer]
The `...` token isn't allowed in the specialization of a class parameterized by a TypeVarTuple. It's allowed only for the `tuple` class as discussed [here](https://typing.readthedocs.io/en/latest/spec/tuples.html#tuple-type-form) in the typing spec.

You can use an unpacked tuple to specialize a class parameterized with a TypeVarTuple, like this:

```python
x: Shape[*tuple[int, ...]]
```

For more information, refer to [the TypeVarTuple section](https://typing.readthedocs.io/en/latest/spec/generics.html#unpacking-unbounded-tuple-types) of the typing spec.

**Reply by MilanStaffehl** [reply]
Thanks, the second link clears up a lot of confusion for me!
